DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,342 @@
+// clang-format off
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+// clang-format on
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+
+// Forward declare the kernel name in the global scope.
+// This is a FPGA best practice that reduces name mangling in the optimization
+// reports.
+class Add;
+class Div;
+class Mult;
+class ShiftLeft;
+class EfficientShiftLeft;
+class GetBitSlice;
+class SetBitSlice;
+
+using ac_int4 = ac_intN::int4;
+using ac_int14 = ac_intN::int14;
+using ac_int15 = ac_intN::int15;
+using ac_int28 = ac_intN::int28;
+using ac_uint4 = ac_intN::uint4;
+
+constexpr int thirteen_bits_mask = (1 << 13) - 1;
+constexpr int fourteen_bits_mask = (1 << 14) - 1;
+
+void TestAdd(queue &q, const ac_int14 &a, const ac_int14 &b, ac_int15 &c) {
```

Comment:
Do we really need separate kernels for each operation?  That results in a lot of extra boiler-plate code, and I'm not sure how it helps demonstrate anything?

Reply:
personally i like separate kernels because otherwise you would get a single kernel with a whole lot of logic. This also makes the reports easier to parse because you can see the shifters independently of the adders, and slices. (I presume... I did not review the reports. I hope they are clear :))